src/mesh/distributed_mesh.C

```diff
@@ -60,14 +60,30 @@ DistributedMesh::DistributedMesh (const Parallel::Communicator & comm_in,
 
 DistributedMesh & DistributedMesh::operator= (DistributedMesh && other_mesh)
 {
-  // First set this mesh equal to other_mesh as an UnstructuredMesh.
+  // First make a call to MeshBase::clear
+  this->MeshBase::clear();
+
+  // Move assign as a ParallelObject.
+  this->ParallelObject::operator=(other_mesh);
+
+  // Move assign as a MeshBase.
+  this->MeshBase::operator=(std::move(other_mesh));
+
+  // Nodes and elements belong to DistributedMesh and have to be
+  // moved before we can move the BoundaryInfo object in UnstructuredMesh::operator=.
+  this->move_nodes_and_elements(std::move(other_mesh));
+
+  // Move assign as an UnstructuredMesh.
   this->UnstructuredMesh::operator=(std::move(other_mesh));
 
+  // The remaining work to move a DistributedMesh
   _is_serial = std::move(reinterpret_cast<DistributedMesh &>(other_mesh).is_serial());
```

Comment:
I should have noticed before, but: std::move() here is pointless.  `is_serial()` returns by value, so it's already a temporary, so it's already an rvalue even without std::move().  And even if it wasn't, there's no move optimization possible with a simple bool; copying a bool is already as cheap as it gets.  So we might as well just set `_is_serial = other_mesh.is_serial()` here and make the code cleaner.

Reply:
Likewise for _is_serial_on_proc_0, _max_node_id, _max_elem_id.

Reply:
Etcetera.  Basically anything that's not doing a heap allocation anywhere under the hood (i.e. anything that's not a container or a complicated class) can just be copied; move is helpful by letting us just move pointers rather than their contents.